### PR TITLE
[WIP]: first prototype of cumulative reduce. functionality is there, but no …

### PR DESCRIPTION
**Pull request by aanthony1243**
…params to change back to regular mode

### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

**Comment by sanderson** [comment]
To toggle this added functionality and output all rows from the reduce operation, we should add a new boolean parameter `emitAll` (or `outputAll` _(never a new feature without some bike-shedding 😄)_).

With `emitAll: true`, output tables include rows for each corresponding input row. The default value is `false` to preserve backwards compatibility with the existing `reduce` behavior.

**Comment by aanthony1243** [comment]
I don't have a strong opinion on the parameter name, i'm not sure if there's precedent in another function that has a similar option, @sanderson maybe you have some idea if that's possibly true?

**Comment by aanthony1243** [comment]
pushed a new version, this is the current behavior: 

```
import "sampledata"

stateChangeTest = (tables=<-, c) => tables
  |> reduce(identity: {prev: 0, current: 0, different: false}, fn: (r, accumulator) => ({ prev: accumulator.current, current: r._value, different: accumulator.current != r._value }), cumulative:c)

sampledata.int() |> stateChangeTest(c: false) |> yield(name: "legacy")
sampledata.int() |> stateChangeTest(c: true) |> yield(name: "cumulative")

----------------------------------
Result: legacy
Table: keys: [tag]
            tag:string                 current:int  different:bool                    prev:int
----------------------  --------------------------  --------------  --------------------------
                    t1                           4            true                          15
Table: keys: [tag]
            tag:string                 current:int  different:bool                    prev:int
----------------------  --------------------------  --------------  --------------------------
                    t2                           1            true                          13
Result: cumulative
Table: keys: [tag]
            tag:string                 current:int  different:bool                    prev:int
----------------------  --------------------------  --------------  --------------------------
                    t1                          -2            true                           0
                    t1                          10            true                          -2
                    t1                           7            true                          10
                    t1                          17            true                           7
                    t1                          15            true                          17
                    t1                           4            true                          15
Table: keys: [tag]
            tag:string                 current:int  different:bool                    prev:int
----------------------  --------------------------  --------------  --------------------------
                    t2                          19            true                           0
                    t2                           4            true                          19
                    t2                          -3            true                           4
                    t2                          19            true                          -3
                    t2                          13            true                          19
                    t2                           1            true                          13
```

**Comment by sanderson** [comment]
@aanthony1243 I don't have a strong opinion either about the parameter name, although I don't know that `cumulative` is the right one. Cumulative calculations are a use case, but I don't think they are _the_ use case. I think `emitAll` or `emitEach` would work.

**Comment by aanthony1243** [comment]
I will go with emitAll

**Comment by lazarenkovegor** [comment]
really looking forward to this feature. Can you tell me if it is possible to collect this feature before it appears in the release? Or is there an instruction on how to build and test your own function in golang? I found only instructions on where to push my functions, but I did not find instructions on how to raise the test environment locally.

**Comment by aanthony1243** [comment]
@lazarenkovegor it's possible to build+test this branch, instructions are in the README.  After you can build flux standalone, linking it into and building influxdb is a straightforward redirect of the go.mod dependency to the local flux directory.

**Comment by geropl** [comment]
I happened to find this PR from this discussion: https://github.com/influxdata/flux/issues/2452#issuecomment-1007382967

Really looking forward for this to get merged, thx for working on it! :pray: 

FYI here's a slightly different (more bulky?) approach to the problem: https://github.com/influxdata/flux/pull/4395
Sharing here in case it might help learn about the design space. :+1:

File: stdlib/universe/reduce.go [hunk 1]
```diff
@@ -151,60 +162,112 @@ func (t *reduceTransformation) Process(id execute.DatasetID, tbl flux.Table) err
 				return errors.Wrap(err, codes.Inherit, "failed to evaluate reduce function")
 			}
 			params[accumulatorParamName] = m
+
+			key := t.computeGroupKey(tbl.Key(), m)
```
Comment (Marwes) [inline stdlib/universe/reduce.go:166]: This seems entirely duplicated from the same code below. Should factor it out to a function if it does not make sense to keep it in the loop (as the added comment mentions).

**Review by Marwes — CHANGES_REQUESTED** [review]

**Pull request closed without merging**